listings/applications/advanced_factory/src/campaign.cairo

```diff
@@ -0,0 +1,286 @@
+            self.ownable._assert_only_owner();
+            assert(self._is_active(), Errors::ENDED);
+            assert(self._is_target_reached(), Errors::TARGET_NOT_REACHED);
+
+            let this = get_contract_address();
+            let token = self.token.read();
+
+            let amount = token.balance_of(this);
+            assert(amount > 0, Errors::ZERO_FUNDS);
+
+            self.status.write(Status::SUCCESSFUL);
+
+            // no need to set total_contributions to 0, as the campaign has ended
+            // and the field can be used as a testament to how much was raised
+
+            let success = token.transfer(get_caller_address(), amount);
+            assert(success, Errors::TRANSFER_FAILED);
+
+            self.emit(Event::Claimed(Claimed { amount }));
+        }
+
+        fn close(ref self: ContractState, reason: ByteArray) {
+            self.ownable._assert_only_owner();
+            assert(self._is_active(), Errors::ENDED);
+
+            self.status.write(Status::CLOSED);
+
+            let mut contributions = self.get_contributions();
+            while let Option::Some((contributor, amt)) = contributions
+                .pop_front() {
+                    self.contributions.remove(contributor);
+                    self.token.read().transfer(contributor, amt);
+                };
+
+            self.emit(Event::Closed(Closed { reason }));
+        }
+
+        fn contribute(ref self: ContractState, amount: u256) {
+            assert(self._is_active(), Errors::ENDED);
+            assert(amount > 0, Errors::ZERO_DONATION);
+
+            let contributor = get_caller_address();
+            let this = get_contract_address();
+            let success = self.token.read().transfer_from(contributor, this, amount.into());
+            assert(success, Errors::TRANSFER_FAILED);
+
+            self.contributions.add(contributor, amount);
+            self.total_contributions.write(self.total_contributions.read() + amount);
+
+            self.emit(Event::ContributionMade(ContributionMade { contributor, amount }));
+        }
+
+        fn get_contributions(self: @ContractState) -> Array<(ContractAddress, u256)> {
+            self.contributions.get_contributions_as_arr()
+        }
+
+        fn get_details(self: @ContractState) -> Details {
+            Details {
+                title: self.title.read(),
+                description: self.description.read(),
+                target: self.target.read(),
+                end_time: self.end_time.read(),
+                status: self.status.read(),
+                token: self.token.read().contract_address,
+                total_contributions: self.total_contributions.read(),
+            }
+        }
+
+        fn upgrade(ref self: ContractState, impl_hash: ClassHash) -> Result<(), Array<felt252>> {
+            if get_caller_address() != self.factory.read() {
+                return Result::Err(array![Errors::NOT_FACTORY]);
+            }
```

Comment:
Maybe it's best to not mention any factory in this contract, and instead rely on owner (or another additional 'manager')